## Fly Emirates data Cleaning

In [1]:
import pandas as pd
import numpy as np

# Load the messy data
flights = pd.read_csv("flights.csv")
airlines = pd.read_csv("airlines.csv")
airports = pd.read_csv("airports.csv")

C:\Users\user\AppData\Local\Temp\ipykernel_12584\3466573493.py:5: DtypeWarning: Columns (7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  flights = pd.read_csv("flights.csv")


In [2]:
print(flights.shape)
print(airlines.shape)
print(airports.shape)

(5819079, 31)
(14, 2)
(322, 7)


In [3]:
flights_clean = flights.copy()

In [4]:
flights_clean.columns

Index(['YEAR', 'MONTH', 'DAY', 'DAY_OF_WEEK', 'AIRLINE', 'FLIGHT_NUMBER',
       'TAIL_NUMBER', 'ORIGIN_AIRPORT', 'DESTINATION_AIRPORT',
       'SCHEDULED_DEPARTURE', 'DEPARTURE_TIME', 'DEPARTURE_DELAY', 'TAXI_OUT',
       'WHEELS_OFF', 'SCHEDULED_TIME', 'ELAPSED_TIME', 'AIR_TIME', 'DISTANCE',
       'WHEELS_ON', 'TAXI_IN', 'SCHEDULED_ARRIVAL', 'ARRIVAL_TIME',
       'ARRIVAL_DELAY', 'DIVERTED', 'CANCELLED', 'CANCELLATION_REASON',
       'AIR_SYSTEM_DELAY', 'SECURITY_DELAY', 'AIRLINE_DELAY',
       'LATE_AIRCRAFT_DELAY', 'WEATHER_DELAY'],
      dtype='object')

In [5]:
flights_clean.columns = (
    flights_clean.columns
    .str.strip()
    .str.lower()
)

In [6]:
flights_clean.columns

Index(['year', 'month', 'day', 'day_of_week', 'airline', 'flight_number',
       'tail_number', 'origin_airport', 'destination_airport',
       'scheduled_departure', 'departure_time', 'departure_delay', 'taxi_out',
       'wheels_off', 'scheduled_time', 'elapsed_time', 'air_time', 'distance',
       'wheels_on', 'taxi_in', 'scheduled_arrival', 'arrival_time',
       'arrival_delay', 'diverted', 'cancelled', 'cancellation_reason',
       'air_system_delay', 'security_delay', 'airline_delay',
       'late_aircraft_delay', 'weather_delay'],
      dtype='object')

In [7]:
flights_clean.describe()

,year,month,day,day_of_week,flight_number,scheduled_departure,departure_time,departure_delay,taxi_out,wheels_off,...,scheduled_arrival,arrival_time,arrival_delay,diverted,cancelled,air_system_delay,security_delay,airline_delay,late_aircraft_delay,weather_delay
count,5819079.0,5.819079e+06,5.819079e+06,5.819079e+06,5.819079e+06,5.819079e+06,5.732926e+06,5.732926e+06,5.730032e+06,5.730032e+06,...,5.819079e+06,5.726566e+06,5.714008e+06,5.819079e+06,5.819079e+06,1.063439e+06,1.063439e+06,1.063439e+06,1.063439e+06,1.063439e+06
mean,2015.0,6.524085e+00,1.570459e+01,3.926941e+00,2.173093e+03,1.329602e+03,1.335204e+03,9.370158e+00,1.607166e+01,1.357171e+03,...,1.493808e+03,1.476491e+03,4.407057e+00,2.609863e-03,1.544643e-02,1.348057e+01,7.615387e-02,1.896955e+01,2.347284e+01,2.915290e+00
std,0.0,3.405137e+00,8.783425e+00,1.988845e+00,1.757064e+03,4.837518e+02,4.964233e+02,3.708094e+01,8.895574e+00,4.980094e+02,...,5.071647e+02,5.263197e+02,3.927130e+01,5.102012e-02,1.233201e-01,2.800368e+01,2.143460e+00,4.816164e+01,4.319702e+01,2.043334e+01
min,2015.0,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,-8.200000e+01,1.000000e+00,1.000000e+00,...,1.000000e+00,1.000000e+00,-8.700000e+01,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,2015.0,4.000000e+00,8.000000e+00,2.000000e+00,7.300000e+02,9.170000e+02,9.210000e+02,-5.000000e+00,1.100000e+01,9.350000e+02,...,1.110000e+03,1.059000e+03,-1.300000e+01,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
50%,2015.0,7.000000e+00,1.600000e+01,4.000000e+00,1.690000e+03,1.325000e+03,1.330000e+03,-2.000000e+00,1.400000e+01,1.343000e+03,...,1.520000e+03,1.512000e+03,-5.000000e+00,0.000000e+00,0.000000e+00,2.000000e+00,0.000000e+00,2.000000e+00,3.000000e+00,0.000000e+00
75%,2015.0,9.000000e+00,2.300000e+01,6.000000e+00,3.230000e+03,1.730000e+03,1.740000e+03,7.000000e+00,1.900000e+01,1.754000e+03,...,1.918000e+03,1.917000e+03,8.000000e+00,0.000000e+00,0.000000e+00,1.800000e+01,0.000000e+00,1.900000e+01,2.900000e+01,0.000000e+00
max,2015.0,1.200000e+01,3.100000e+01,7.000000e+00,9.855000e+03,2.359000e+03,2.400000e+03,1.988000e+03,2.250000e+02,2.400000e+03,...,2.400000e+03,2.400000e+03,1.971000e+03,1.000000e+00,1.000000e+00,1.134000e+03,5.730000e+02,1.971000e+03,1.331000e+03,1.211000e+03


In [ ]:
## check for missing values

In [8]:
missing = flights_clean.isnull().sum()

missing[missing > 0].sort_values(ascending=False)

cancellation_reason    5729195
weather_delay          4755640
late_aircraft_delay    4755640
airline_delay          4755640
security_delay         4755640
air_system_delay       4755640
elapsed_time            105071
air_time                105071
arrival_delay           105071
arrival_time             92513
taxi_in                  92513
wheels_on                92513
wheels_off               89047
taxi_out                 89047
departure_time           86153
departure_delay          86153
tail_number              14721
scheduled_time               6
dtype: int64

In [9]:
missing_percent = (
    flights_clean.isnull().sum()
    / len(flights_clean)
) * 100

missing_percent.sort_values(ascending=False)

cancellation_reason    98.455357
weather_delay          81.724960
late_aircraft_delay    81.724960
airline_delay          81.724960
security_delay         81.724960
air_system_delay       81.724960
air_time                1.805629
arrival_delay           1.805629
elapsed_time            1.805629
wheels_on               1.589822
taxi_in                 1.589822
arrival_time            1.589822
taxi_out                1.530259
wheels_off              1.530259
departure_delay         1.480526
departure_time          1.480526
tail_number             0.252978
scheduled_time          0.000103
scheduled_departure     0.000000
cancelled               0.000000
day                     0.000000
day_of_week             0.000000
airline                 0.000000
flight_number           0.000000
scheduled_arrival       0.000000
diverted                0.000000
origin_airport          0.000000
distance                0.000000
destination_airport     0.000000
month                   0.000000
year      

In [15]:
flights_clean[
    flights_clean['cancelled'] == 1
].head()

,year,month,day,day_of_week,airline,flight_number,tail_number,origin_airport,destination_airport,scheduled_departure,...,arrival_time,arrival_delay,diverted,cancelled,cancellation_reason,air_system_delay,security_delay,airline_delay,late_aircraft_delay,weather_delay
32,2015,1,1,4,AS,136,N431AS,ANC,SEA,135,...,NaN,NaN,0,1,A,NaN,NaN,NaN,NaN,NaN
42,2015,1,1,4,AA,2459,N3BDAA,PHX,DFW,200,...,NaN,NaN,0,1,B,NaN,NaN,NaN,NaN,NaN
68,2015,1,1,4,OO,5254,N746SK,MAF,IAH,510,...,NaN,NaN,0,1,B,NaN,NaN,NaN,NaN,NaN
82,2015,1,1,4,MQ,2859,N660MQ,SGF,DFW,525,...,NaN,NaN,0,1,B,NaN,NaN,NaN,NaN,NaN
90,2015,1,1,4,OO,5460,N583SW,RDD,SFO,530,...,NaN,NaN,0,1,A,NaN,NaN,NaN,NaN,NaN



arrival_time     = NaN
arrival_delay    = NaN

for cancelled flights.

This tells us the missing values are not data quality issues.

They're missing because:

Flight was cancelled
→ Flight never arrived
→ No arrival time exists
→ No arrival delay exists

In [16]:
missing = flights_clean.isnull().sum()

missing[missing > 0].sort_values(ascending=False)

cancellation_reason    5729195
weather_delay          4755640
late_aircraft_delay    4755640
airline_delay          4755640
security_delay         4755640
air_system_delay       4755640
elapsed_time            105071
air_time                105071
arrival_delay           105071
arrival_time             92513
taxi_in                  92513
wheels_on                92513
wheels_off               89047
taxi_out                 89047
departure_time           86153
departure_delay          86153
tail_number              14721
scheduled_time               6
dtype: int64

In [17]:
flights_clean.groupby('cancelled')['cancellation_reason'].apply(
    lambda x: x.isnull().sum()
)

cancelled
0    5729195
1          0
Name: cancellation_reason, dtype: int64

In [18]:
pd.crosstab(
    flights_clean['cancelled'],
    flights_clean['cancellation_reason'].isnull()
)

cancellation_reason,False,True
cancelled,,
0,0,5729195
1,89884,0


For analysis and Power BI readability, I would replace NULL/0 with a descriptive label:'Not Cancelled'

In [19]:
flights_clean['cancellation_reason'] = flights_clean[
    'cancellation_reason'
].fillna('Not Cancelled')

In [20]:
flights_clean['cancellation_reason'].isnull().sum()

np.int64(0)

In [22]:
flights_clean[['weather_delay']].describe()

,weather_delay
count,1.063439e+06
mean,2.915290e+00
std,2.043334e+01
min,0.000000e+00
25%,0.000000e+00
50%,0.000000e+00
75%,0.000000e+00
max,1.211000e+03


In [23]:
flights_clean['weather_delay'].value_counts().head(20)

weather_delay
0.0     998723
6.0       1649
8.0       1580
7.0       1537
10.0      1498
15.0      1498
9.0       1487
16.0      1460
5.0       1415
3.0       1412
2.0       1397
11.0      1383
17.0      1378
4.0       1319
12.0      1312
1.0       1308
13.0      1271
18.0      1230
14.0      1172
19.0      1163
Name: count, dtype: int64

Fill missing values with 0

In [24]:
flights_clean['weather_delay'] = flights_clean[
    'weather_delay'
].fillna(0)

In [25]:
flights_clean['weather_delay'].isnull().sum()

np.int64(0)

In [26]:
missing = flights_clean.isnull().sum()

missing[missing > 0].sort_values(ascending=False)

air_system_delay       4755640
security_delay         4755640
airline_delay          4755640
late_aircraft_delay    4755640
elapsed_time            105071
air_time                105071
arrival_delay           105071
wheels_on                92513
taxi_in                  92513
arrival_time             92513
taxi_out                 89047
wheels_off               89047
departure_time           86153
departure_delay          86153
tail_number              14721
scheduled_time               6
dtype: int64

late_aircraft_delay

In [27]:
flights_clean['late_aircraft_delay'].describe()

count    1.063439e+06
mean     2.347284e+01
std      4.319702e+01
min      0.000000e+00
25%      0.000000e+00
50%      3.000000e+00
75%      2.900000e+01
max      1.331000e+03
Name: late_aircraft_delay, dtype: float64

In [28]:
flights_clean['late_aircraft_delay'].value_counts().head(10)

late_aircraft_delay
0.0     506486
15.0     14522
16.0     13824
17.0     12908
18.0     12259
19.0     11794
14.0     11183
20.0     11079
13.0     10930
11.0     10517
Name: count, dtype: int64

This tells us:
0 = Flight was not delayed because of a late arriving aircraft
Positive number = Delay minutes caused by a late arriving aircraft
NULL = No late-aircraft delay information recorded

In [29]:
flights_clean['late_aircraft_delay'] = flights_clean[
    'late_aircraft_delay'
].fillna(0)

In [30]:
flights_clean['late_aircraft_delay'].isnull().sum()

np.int64(0)

airline_delay

In [31]:
flights_clean['airline_delay'].describe()

count    1.063439e+06
mean     1.896955e+01
std      4.816164e+01
min      0.000000e+00
25%      0.000000e+00
50%      2.000000e+00
75%      1.900000e+01
max      1.971000e+03
Name: airline_delay, dtype: float64

In [32]:
flights_clean['airline_delay'].value_counts().head(10)

airline_delay
0.0     493417
1.0      21319
2.0      21211
3.0      20656
4.0      20184
6.0      20107
5.0      19772
7.0      18646
8.0      17494
15.0     16582
Name: count, dtype: int64

In [34]:
flights_clean['airline_delay'] = flights_clean['airline_delay'].fillna(0)

In [35]:
flights_clean['airline_delay'].isnull().sum()

np.int64(0)

In [36]:
flights_clean['security_delay'].value_counts().head(10)

security_delay
0.0     1059955
15.0        158
8.0         127
10.0        125
12.0        124
6.0         121
7.0         119
13.0        119
9.0         117
5.0         116
Name: count, dtype: int64

In [37]:
flights_clean['security_delay'] = flights_clean['security_delay'].fillna(0)

In [38]:
flights_clean['security_delay'].isnull().sum()

np.int64(0)

In [40]:
flights_clean['air_system_delay'] = flights_clean['security_delay'].fillna(0)

In [41]:
flights_clean['air_system_delay'].isnull().sum()

np.int64(0)

In [42]:
missing = flights_clean.isnull().sum()

missing[missing > 0].sort_values(ascending=False)

elapsed_time       105071
air_time           105071
arrival_delay      105071
wheels_on           92513
taxi_in             92513
arrival_time        92513
taxi_out            89047
wheels_off          89047
departure_time      86153
departure_delay     86153
tail_number         14721
scheduled_time          6
dtype: int64

elapsed_time

In [43]:
flights_clean[
    flights_clean['elapsed_time'].isnull()
]['cancelled'].value_counts()

cancelled
1    89884
0    15187
Name: count, dtype: int64

for 89884 Cancelled Flight
→ Never completed journey
→ No elapsed time exists

for 15,187 we will check :

In [44]:
flights_clean[
    (flights_clean['cancelled'] == 0) &
    (flights_clean['elapsed_time'].isnull())
][[
    'flight_number',
    'origin_airport',
    'destination_airport',
    'departure_time',
    'arrival_time',
    'air_time',
    'elapsed_time',
    'diverted'
]].head(10)

,flight_number,origin_airport,destination_airport,departure_time,arrival_time,air_time,elapsed_time,diverted
724,716,OMA,ATL,637.0,1451.0,NaN,NaN,1
1455,5237,MKE,IAH,742.0,1505.0,NaN,NaN,1
2218,1966,ATL,JAX,922.0,1219.0,NaN,NaN,1
2526,4555,IAH,HRL,859.0,1319.0,NaN,NaN,1
2866,1081,MDW,OKC,926.0,1343.0,NaN,NaN,1
4225,1279,OMA,DFW,1058.0,1838.0,NaN,NaN,1
4394,4506,SLC,SUN,1108.0,1353.0,NaN,NaN,1
4574,6171,DEN,GUC,1120.0,1635.0,NaN,NaN,1
4636,4169,IAH,ABQ,1122.0,1630.0,NaN,NaN,1
4639,4654,IAH,HRL,1206.0,NaN,NaN,NaN,1


In [45]:
flights_clean[
    (flights_clean['cancelled'] == 0) &
    (flights_clean['elapsed_time'].isnull())
]['diverted'].value_counts()


diverted
1    15187
Name: count, dtype: int64

Missing elapsed_time
→ Mostly diverted flights
→ Not dirty data

we will keep them null because if we fill 0, mean(), median() would create fake flight durations.

In [46]:
flights_clean[
    flights_clean['air_time'].isnull()
]['cancelled'].value_counts()

cancelled
1    89884
0    15187
Name: count, dtype: int64

In [48]:
flights_clean[
    flights_clean['arrival_delay'].isnull()
]['cancelled'].value_counts()

cancelled
1    89884
0    15187
Name: count, dtype: int64

In [47]:
flights_clean[
    flights_clean['arrival_time'].isnull()
]['cancelled'].value_counts()

cancelled
1    89884
0     2629
Name: count, dtype: int64

In [ ]:
Cancelled flight
→ Never completed journey
→ No arrival time
→ No arrival delay
→ No air time
→ No elapsed time

Diverted flight
→ Original route not completed normally
→ Some operational metrics unavailable

In [49]:
remaining_cols = [
    'taxi_in',
    'wheels_on',
    'wheels_off',
    'taxi_out',
    'departure_time',
    'departure_delay',
    'tail_number',
    'scheduled_time'
]

for col in remaining_cols:
    print(f"\n----- {col} -----")
    print(
        flights_clean[
            flights_clean[col].isnull()
        ][['cancelled', 'diverted']]
        .value_counts()
        .head()
    )


----- taxi_in -----
cancelled  diverted
1          0           89884
0          1            2629
Name: count, dtype: int64

----- wheels_on -----
cancelled  diverted
1          0           89884
0          1            2629
Name: count, dtype: int64

----- wheels_off -----
cancelled  diverted
1          0           89047
Name: count, dtype: int64

----- taxi_out -----
cancelled  diverted
1          0           89047
Name: count, dtype: int64

----- departure_time -----
cancelled  diverted
1          0           86153
Name: count, dtype: int64

----- departure_delay -----
cancelled  diverted
1          0           86153
Name: count, dtype: int64

----- tail_number -----
cancelled  diverted
1          0           14721
Name: count, dtype: int64

----- scheduled_time -----
cancelled  diverted
1          0           5
0          1           1
Name: count, dtype: int64


In [50]:
#labelling null with unknown in tail number because it indicates aircraft not assigned.
flights_clean['tail_number'] = flights_clean[
    'tail_number'
].fillna('Unknown')

In [52]:
flights_clean['scheduled_time'].isnull().sum()

np.int64(6)

only 6 value are null in scheduled time which is 0.0001% that why I remove them. 

In [53]:
flights_clean = flights_clean.dropna(
    subset=['scheduled_time']
)

In [54]:
flights_clean['scheduled_time'].isnull().sum()

np.int64(0)

In [55]:
flights_clean.isnull().sum().sort_values(ascending=False)

elapsed_time           105065
air_time               105065
arrival_delay          105065
wheels_on               92507
arrival_time            92507
taxi_in                 92507
taxi_out                89047
wheels_off              89047
departure_time          86153
departure_delay         86153
airline_delay               0
security_delay              0
air_system_delay            0
cancellation_reason         0
cancelled                   0
diverted                    0
scheduled_arrival           0
late_aircraft_delay         0
year                        0
distance                    0
month                       0
scheduled_time              0
scheduled_departure         0
destination_airport         0
origin_airport              0
tail_number                 0
flight_number               0
airline                     0
day_of_week                 0
day                         0
weather_delay               0
dtype: int64

Missing values were investigated individually. Delay-cause columns (weather_delay, airline_delay, security_delay, air_system_delay, late_aircraft_delay) were filled with 0 because NULL represented the absence of that delay type. Cancellation_reason NULLs were replaced with "Not Cancelled". Tail_number NULLs were replaced with "Unknown". Operational flight columns such as departure_time, arrival_time, air_time, and elapsed_time were retained as NULL because they corresponded to cancelled or diverted flights and represent valid business scenarios.

In [56]:
flights_clean.duplicated().sum()

np.int64(0)

datatype overview

In [57]:
flights_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5819073 entries, 0 to 5819078
Data columns (total 31 columns):
 #   Column               Dtype  
---  ------               -----  
 0   year                 int64  
 1   month                int64  
 2   day                  int64  
 3   day_of_week          int64  
 4   airline              object 
 5   flight_number        int64  
 6   tail_number          object 
 7   origin_airport       object 
 8   destination_airport  object 
 9   scheduled_departure  int64  
 10  departure_time       float64
 11  departure_delay      float64
 12  taxi_out             float64
 13  wheels_off           float64
 14  scheduled_time       float64
 15  elapsed_time         float64
 16  air_time             float64
 17  distance             int64  
 18  wheels_on            float64
 19  taxi_in              float64
 20  scheduled_arrival    int64  
 21  arrival_time         float64
 22  arrival_delay        float64
 23  diverted             int64  
 24  can

Create necesary columns 

In [58]:
flights_clean['flight_date'] = pd.to_datetime(
    flights_clean[['year', 'month', 'day']]
)

C:\Users\user\AppData\Local\Temp\ipykernel_12584\2847066470.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  flights_clean['flight_date'] = pd.to_datetime(


In [59]:
flights_clean[['year','month','day','flight_date']].head()

,year,month,day,flight_date
0,2015,1,1,2015-01-01
1,2015,1,1,2015-01-01
2,2015,1,1,2015-01-01
3,2015,1,1,2015-01-01
4,2015,1,1,2015-01-01


In [61]:
flights_clean['month_name'] = flights_clean['flight_date'].dt.month_name()

C:\Users\user\AppData\Local\Temp\ipykernel_12584\1932222442.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  flights_clean['month_name'] = flights_clean['flight_date'].dt.month_name()


In [64]:
flights_clean['day_name'] = flights_clean['flight_date'].dt.day_name()

C:\Users\user\AppData\Local\Temp\ipykernel_12584\3948369448.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  flights_clean['day_name'] = flights_clean['flight_date'].dt.day_name()


In [65]:
flights_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5819073 entries, 0 to 5819078
Data columns (total 34 columns):
 #   Column               Dtype         
---  ------               -----         
 0   year                 int64         
 1   month                int64         
 2   day                  int64         
 3   day_of_week          int64         
 4   airline              object        
 5   flight_number        int64         
 6   tail_number          object        
 7   origin_airport       object        
 8   destination_airport  object        
 9   scheduled_departure  int64         
 10  departure_time       float64       
 11  departure_delay      float64       
 12  taxi_out             float64       
 13  wheels_off           float64       
 14  scheduled_time       float64       
 15  elapsed_time         float64       
 16  air_time             float64       
 17  distance             int64         
 18  wheels_on            float64       
 19  taxi_in              float

2. lets do for airlines tables 

In [69]:
airlines_clean = airlines.copy()

In [74]:
airlines_clean.columns

Index(['IATA_CODE', 'AIRLINE'], dtype='object')

In [75]:
airlines_clean.columns = (
    airlines_clean.columns
    .str.strip()
    .str.lower()
)

In [76]:
airlines_clean.columns

Index(['iata_code', 'airline'], dtype='object')

In [77]:
airlines_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   iata_code  14 non-null     object
 1   airline    14 non-null     object
dtypes: object(2)
memory usage: 356.0+ bytes


In [78]:
airlines_clean.isnull().sum()

iata_code    0
airline      0
dtype: int64

In [79]:
airlines_clean.duplicated().sum()

np.int64(0)

In [80]:
airlines_clean.head()

,iata_code,airline
0,UA,United Air Lines Inc.
1,AA,American Airlines Inc.
2,US,US Airways Inc.
3,F9,Frontier Airlines Inc.
4,B6,JetBlue Airways


3. for airports table

In [81]:
airports_clean = airports.copy()

In [82]:
airports_clean.columns

Index(['IATA_CODE', 'AIRPORT', 'CITY', 'STATE', 'COUNTRY', 'LATITUDE',
       'LONGITUDE'],
      dtype='object')

In [83]:
airports_clean.columns = (
    airports_clean.columns
    .str.strip()
    .str.lower()
)

In [84]:
airports_clean.columns

Index(['iata_code', 'airport', 'city', 'state', 'country', 'latitude',
       'longitude'],
      dtype='object')

In [86]:
airports_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322 entries, 0 to 321
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   iata_code  322 non-null    object 
 1   airport    322 non-null    object 
 2   city       322 non-null    object 
 3   state      322 non-null    object 
 4   country    322 non-null    object 
 5   latitude   319 non-null    float64
 6   longitude  319 non-null    float64
dtypes: float64(2), object(5)
memory usage: 17.7+ KB


In [85]:
airports_clean.isnull().sum()

iata_code    0
airport      0
city         0
state        0
country      0
latitude     3
longitude    3
dtype: int64

In [87]:
airports_clean[
    airports_clean['latitude'].isnull()
]

,iata_code,airport,city,state,country,latitude,longitude
96,ECP,Northwest Florida Beaches International Airport,Panama City,FL,USA,NaN,NaN
234,PBG,Plattsburgh International Airport,Plattsburgh,NY,USA,NaN,NaN
313,UST,Northeast Florida Regional Airport (St. August...,St. Augustine,FL,USA,NaN,NaN


In [88]:
airports_clean[
    airports_clean['longitude'].isnull()
]

,iata_code,airport,city,state,country,latitude,longitude
96,ECP,Northwest Florida Beaches International Airport,Panama City,FL,USA,NaN,NaN
234,PBG,Plattsburgh International Airport,Plattsburgh,NY,USA,NaN,NaN
313,UST,Northeast Florida Regional Airport (St. August...,St. Augustine,FL,USA,NaN,NaN


In [89]:
airports_clean.duplicated().sum()

np.int64(0)

In [90]:
airports_clean['iata_code'].duplicated().sum()

np.int64(0)

In [91]:
flights_clean.to_csv("flights_clean.csv", index=False)
airlines_clean.to_csv("airlines_clean.csv", index=False)
airports_clean.to_csv("airports_clean.csv", index=False)

In [93]:
flights_clean.memory_usage(deep=True).sum() / 1024**2

np.float64(3354.874997138977)